cairo-libs/packages/protobuf/src/tests/oneof.cairo

```diff
@@ -0,0 +1,127 @@
+use protobuf::types::message::{
+    ProtoMessage, ProtoOneof, DecodeContext, EncodeContext, EncodeContextTrait, DecodeContextTrait,
+    ProtoCodecImpl,
+};
+use protobuf::types::tag::{WireType, ProtobufTag};
+use protobuf::primitives::numeric::U64AsProtoMessage;
+use protobuf::primitives::array::ByteArrayAsProtoMessage;
+
+#[derive(Drop, Debug, Default, PartialEq)]
+pub struct MessageWithOneof {
+    oneof: Oneof,
+}
+
+pub impl MessageWithOneofAsProtoMessage of ProtoMessage<MessageWithOneof> {
+    fn encode_raw(self: @MessageWithOneof, ref context: EncodeContext) {
+        context.encode_oneof(self.oneof);
+    }
+
+    fn decode_raw(ref context: DecodeContext) -> Option<MessageWithOneof> {
+        let oneof = context.decode_oneof()?;
+        Option::Some(MessageWithOneof { oneof })
+    }
+
+    fn wire_type() -> WireType {
+        WireType::LengthDelimited
+    }
+}
+
+#[derive(Drop, Debug, Default, PartialEq)]
+pub enum Oneof {
+    #[default]
+    Int: u64,
+    String: ByteArray,
+    Enum: Enum,
+}
+
+pub impl OneofAsProtoOneof of ProtoOneof<Oneof> {
+    fn encode_raw(self: @Oneof, ref context: EncodeContext) -> ProtobufTag {
+        match self {
+            Oneof::Int(n) => {
+                n.encode_raw(ref context);
+                let wire_type = ProtoMessage::<u64>::wire_type();
+                ProtobufTag { field_number: 1, wire_type }
+            },
+            Oneof::String(str) => {
+                str.encode_raw(ref context);
+                let wire_type = ProtoMessage::<ByteArray>::wire_type();
+                ProtobufTag { field_number: 2, wire_type }
+            },
+            Oneof::Enum(e) => {
+                let n: u32 = e.into();
+                n.encode_raw(ref context);
+                let wire_type = ProtoMessage::<u32>::wire_type();
+                ProtobufTag { field_number: 3, wire_type }
+            },
+        }
+    }
+
+    fn decode_raw(ref context: DecodeContext, tag: ProtobufTag) -> Option<Oneof> {
+        if tag.field_number == 1 {
```

Comment:
We couldn't use a `match` statement because it expects patterns to start from 0. Additionally, Cairo does not support `match` statements on tuples like (`field_number`, `wire_type`).